aip/general/0236.md

```diff
@@ -0,0 +1,355 @@
+  string parent_etag = 3;
+  bool successful = 3;
+  string error_message = 4;
+}
+```
+- Admin flow API to start previewing a resource **should** be implemented.
+  - Remains declarative friendly as only an output only field is modified.
+- If startPreview is called specifically on a resource which is already being previewed, then it **should** return an error message.
+- When startPreview is called, the system **should** set values for the log_name, and start_time in PreviewMetadata and set the state to ACTIVE.
+- If startPreview is called on an experiment with an empty set of rules, the deletion of that resource **must** be previewed.
+- startPreview is not idempotent and using it twice on the same resource **should** return an error.
+- If the API has an error, no fields in PreviewMetadata **should** change.
+- Etags **must** be used to guarantee resource freshness.
+
+#### stopPreview
+
+```proto
+rpc StopPreviewPolicyExperiment(StopPreviewPolicyExperimentRequest) returns (StopPreviewPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:stopPreview"
+    body: "*"
+  }
+}
+
+message StopPreviewPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+}
+
+message StopPreviewPolicyExperimentResponse {
+  string name = 1;
+  string etag = 2;
+  string parent_etag = 3;
+  bool successful = 4;
+  string error_message = 5;
+}
+```
+- Admin flow API to stop previewing a resource **should** be implemented.
+  - Remains declarative friendly as only an output only field is modified.
+- If stopPreview is called specifically on a resource which is not actively being previewed, then it **should** return an error message.
+- When the API is called, PreviewMetadata.state **must** be set to INACTIVE and PreviewMetadata.stop_time **should** be set to the current time.
+- stopPreview is not idempotent, calling stop twice on the same resource **should** return an error.
+- If the API has an error, no fields in PreviewMetadata **should** change.
+- Etags **must** be used to guarantee resource freshness.
+
+#### commit
+
+A new API called commit **may** be implemented to promote the experimental configuration of a resource to the effective version and perform necessary cleanup. It will copy the fields-under-test from the experiment into the parent and then delete the experiment.
+
+This is a convenience API which is not declarative friendly. The alternative supported for declarative clients is to manually copy fields from an experiment onto the live policy then delete the experiment.
+
+```proto
+rpc CommitPolicyExperiment(CommitPolicyExperimentRequest) returns (CommitPolicyExperimentResponse) {
+  option (google.api.http) = {
+    post: "/v1/{name=policies/*/policyExperiments/*}:commit"
+    body: "*"
+  }
+}
+
+message CommitPolicyExperimentRequest {
+  string name = 1;
+  string etag = 2;
+  string parent_etag = 3; // optional.
+}
+
+message CommitPolicyExperimentResponse {
```

Comment:
Same note as StartPreviewPolicyExperimentResponse message --- consider returning the PolicyExperiment as, or an LRO-of-PolicyExperiment if the stop/cancel method needs to be long-running.

Also consider adding a `commit_time` to the experiment resource. That would complete the pairing between of standard+custom methods and the resource timestamps, e.g.

create_time
update_time
start_time
cancel_time
commit_time

(or whatever the custom methods are named)

Reply:
Updated the return value but want to still push back on delete_automatically as it defeats the purpose IMO